src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -6257,186 +6257,195 @@ void StorageReplicatedMergeTree::replacePartitionFrom(
+                entry_replace.src_part_names.emplace_back(part->name);
+            for (const auto & part : dst_parts)
+                entry_replace.new_part_names.emplace_back(part->name);
+            for (const String & checksum : part_checksums)
+                entry_replace.part_names_checksums.emplace_back(checksum);
+            entry_replace.columns_version = -1;
+        }
 
-    ReplicatedMergeTreeLogEntryData entry;
-    {
-        auto src_table_id = src_data.getStorageID();
-        entry.type = ReplicatedMergeTreeLogEntryData::REPLACE_RANGE;
-        entry.source_replica = replica_name;
-        entry.create_time = time(nullptr);
-        entry.replace_range_entry = std::make_shared<ReplicatedMergeTreeLogEntryData::ReplaceRangeEntry>();
-
-        auto & entry_replace = *entry.replace_range_entry;
-        entry_replace.drop_range_part_name = drop_range_fake_part_name;
-        entry_replace.from_database = src_table_id.database_name;
-        entry_replace.from_table = src_table_id.table_name;
-        for (const auto & part : src_parts)
-            entry_replace.src_part_names.emplace_back(part->name);
-        for (const auto & part : dst_parts)
-            entry_replace.new_part_names.emplace_back(part->name);
-        for (const String & checksum : part_checksums)
-            entry_replace.part_names_checksums.emplace_back(checksum);
-        entry_replace.columns_version = -1;
-    }
-
-    /// Remove deduplication block_ids of replacing parts
-    if (replace)
-        clearBlocksInPartition(*zookeeper, drop_range.partition_id, drop_range.max_block, drop_range.max_block);
+        /// Remove deduplication block_ids of replacing parts
+        if (replace)
+            clearBlocksInPartition(*zookeeper, drop_range.partition_id, drop_range.max_block, drop_range.max_block);
 
-    DataPartsVector parts_to_remove;
-    Coordination::Responses op_results;
+        DataPartsVector parts_to_remove;
+        Coordination::Responses op_results;
 
-    try
-    {
-        Coordination::Requests ops;
-        for (size_t i = 0; i < dst_parts.size(); ++i)
+        try
         {
-            getCommitPartOps(ops, dst_parts[i], block_id_paths[i]);
-            ephemeral_locks[i].getUnlockOps(ops);
-
-            if (ops.size() > zkutil::MULTI_BATCH_SIZE)
+            Coordination::Requests ops;
+            for (size_t i = 0; i < dst_parts.size(); ++i)
             {
-                /// It is unnecessary to add parts to working set until we commit log entry
-                zookeeper->multi(ops);
-                ops.clear();
+                getCommitPartOps(ops, dst_parts[i], block_id_paths[i]);
+                ephemeral_locks[i].getUnlockOps(ops);
+
+                if (ops.size() > zkutil::MULTI_BATCH_SIZE)
+                {
+                    /// It is unnecessary to add parts to working set until we commit log entry
+                    zookeeper->multi(ops);
+                    ops.clear();
+                }
```

Comment:
I didn't like this logic, because it may be not atomic with creation of the log entry if parts set is large enough. But now, with retries, it become even for fragile. What do you thing about removing it?

Reply:
I guess the code exists because we'd like to support general `ALTER REPLACE/ATTACH PARTITION` statement. I'm not sure if it'll break anything after we commit some garbage to the part sets in zookeeper but don't generate related log entries.

Reply:
I don't know why this code exists, probably because ZooKeeper has a limitation for request size, 1MB by default (it's configurable). However, this code limits number of operations in `multi` request, not the total size in bytes. 

As for retries, if some parts were committed on the first iteration, but after that we got `ZBADVERSION` on attempt to check `alter_partition_version`, then we will try to commit the same parts again on the next iteration and most likely will fail, because node already exists.

Reply:
> However, this code limits number of operations in multi request, not the total size in bytes.

I guess it's because we used to store much more bytes per parts in zookeeper without `use_minimalistic_checksums_in_zookeeper` . I think we should get rid of this code.


>  then we will try to commit the same parts again on the next iteration and most likely will fail, because node already exists.

Indeed.